### ISSUE TITLE
Use `$lib` in tests files too

### ISSUE THREAD

**Issue by frederikhors** [sveltejs/kit#8742]
### Describe the problem

It would be very useful to use `$lib` in playwright tests files too.

```
Error: Cannot find package '$lib' imported from C:\project\src\lib\helpers.ts
```

### Describe the proposed solution

Rich's Magic

### Alternatives considered

Wait for Godot

### Importance

would make my life easier

**Comment by dummdidumm** [comment]
Judging from https://github.com/microsoft/playwright/pull/10525 and https://github.com/microsoft/playwright/issues/16164 this should already exist, not sure why this doesn't work then.

**Comment by WilliamHua** [comment]
https://github.com/WilliamHua/lib-sveltekit-bug-playwright

It works using version `1.1.1` of SvelteKit, but not `1.1.2`. 

I suspect it's because of this change: https://github.com/sveltejs/kit/pull/8568

You can temporarily work around this by modifying your `tsconfig.json` and adding

```
{
	"extends": "./.svelte-kit/tsconfig.json",
	"compilerOptions": {
                ...
		"baseUrl": ".",
		"paths": {
			"$lib": [
				"src/lib"
			],
			"$lib/*": [
				"src/lib/*"
			]
		}
	}
}
```

**Comment by Smirow** [comment]
Same issue `$env`, but can't really use the solution above. Do we have any other way to access `env` in dev?

**Comment by TadasV1** [comment]
> https://github.com/WilliamHua/lib-sveltekit-bug-playwright
> 
> It works using version `1.1.1` of SvelteKit, but not `1.1.2`.
> 
> I suspect it's because of this change: #8568
> 
> You can temporarily work around this by modifying your `tsconfig.json` and adding
> 
> ```
> {
> 	"extends": "./.svelte-kit/tsconfig.json",
> 	"compilerOptions": {
>                 ...
> 		"baseUrl": ".",
> 		"paths": {
> 			"$lib": [
> 				"src/lib"
> 			],
> 			"$lib/*": [
> 				"src/lib/*"
> 			]
> 		}
> 	}
> }
> ```

not working in playwright. still throws cannot find $lib

**Comment by iamthe-Wraith** [comment]
any update on this? trying to use playwrights component testing and getting this same error.

**Comment by iamthe-Wraith** [comment]
has anyone been able to resolve this? even just a hacky work around?

**Comment by s3812497** [comment]
The following workaround does it for me. For whatever reason it only works when the `baseUrl` option is included.
`/tsconfig.json`
```json
{
	"extends": "./.svelte-kit/tsconfig.json",
	"compilerOptions": {
		"allowJs": true,
		"checkJs": true,
		"esModuleInterop": true,
		"forceConsistentCasingInFileNames": true,
		"resolveJsonModule": true,
		"skipLibCheck": true,
		"sourceMap": true,
		"strict": true,
		"baseUrl": ".",
		"paths": {
			"$lib": [
				"./src/lib"
			],
			"$lib/*": [
				"./src/lib/*"
			]
		},
	},
}
```

**Comment by iamthe-Wraith** [comment]
> The following workaround does it for me ([as mentioned by William](https://github.com/sveltejs/kit/issues/8742#issuecomment-1407449856)). For whatever reason it only works when the `baseUrl` option is included. `/tsconfig.json`

This worked for me, thank you!

**Comment by coryvirok** [comment]
I'm running into the same issue with accessing `$env` from tests, just like @Smirow mentioned. Is there a workaround for this?

**Comment by DylanStandridge** [comment]
I'm also running into the same issue within a test.

**Comment by frederikhors** [comment]
@s3812497 @WilliamHua what about this?

`Error: Cannot find package '$app' imported from ...`?

**Comment by s3812497** [comment]
> what about this?
> 
> `Error: Cannot find package '$app' imported from ...`?

See https://github.com/sveltejs/kit/issues/1485

**Comment by frederikhors** [comment]
@s3812497 I don't know what to read for fixing this.

**Comment by s3812497** [comment]
It doesn't seem like there's a straight-forward fix at the moment. There are a few examples in that issue of people mocking the $app module for their testing frameworks.

**Comment by frederikhors** [comment]
I can use:

```json
"$app": [
	"./node_modules/@sveltejs/kit/src/runtime/app"
],
"$app/*": [
	"./node_modules/@sveltejs/kit/src/runtime/app/*"
],
"$env": [
	"./node_modules/@sveltejs/kit/src/runtime/env"
],
"$env/*": [
	"./node_modules/@sveltejs/kit/src/runtime/env/*"
]
```

but for example `$env/static/public` is a problem.

**Comment by coryvirok** [comment]
I'm using this and it seems to work:

```js
// tsconfig.json
{
  "extends": "./.svelte-kit/tsconfig.json",
  "compilerOptions": {
    // Everything below is required for Playwright tests to access path aliases
    "baseUrl": ".",
    "paths": {
      "$lib": ["src/lib"],
      "$lib/*": ["src/lib/*"],
      "$app/*": ["tests/e2e/mocks/app/*"],
      "$env/*": ["tests/e2e/mocks/env/*"]
    }
  }
}
```

```js
// tests/e2e/mocks/env/app/dynamic/private.ts
export const env = process.env
```

**Comment by frederikhors** [comment]
@coryvirok Is there a way to put `tsconfig.json` into `tests/` dir and not in the root one?

**Comment by coryvirok** [comment]
I haven't tried it, but it looks like you'll need to compile separately:

from https://playwright.dev/docs/test-typescript 
```
{
  "scripts": {
    "pretest": "tsc --incremental -p tests/tsconfig.json",
    "test": "playwright test -c tests-out"
  }
}
```

**Comment by frederikhors** [comment]
> I'm using this and it seems to work:
> 
> ```js
> // tsconfig.json
> {
>   "extends": "./.svelte-kit/tsconfig.json",
>   "compilerOptions": {
>     // Everything below is required for Playwright tests to access path aliases
>     "baseUrl": ".",
>     "paths": {
>       "$lib": ["src/lib"],
>       "$lib/*": ["src/lib/*"],
>       "$app/*": ["tests/e2e/mocks/app/*"],
>       "$env/*": ["tests/e2e/mocks/env/*"]
>     }
>   }
> }
> ```
> 
> ```js
> // tests/e2e/mocks/env/app/dynamic/private.ts
> export const env = process.env
> ```

I cannot make this work. Can you please push a repo to github, @coryvirok?

**Comment by coryvirok** [comment]
For sure. Here is the SvelteKit skeleton app with upgraded deps + env mocks for Playwright.

https://github.com/coryvirok/tmp-mock-env-sveltekit

**Comment by coryvirok** [comment]
I'm not familiar with the svelte query lib but here's what I would recommend:

Add another file to your mocks folder called app.ts. From there, export whichever vars are used by your app at build and runtime. I do this for the `building` var. Then, rename the path aliases in tsconfig for $app to point to this new mock, just like the $env var does.

Your baseUrl looks correct.

**Comment by sowdowdow** [comment]
Hello [same issue here](https://stackoverflow.com/questions/77314398/how-to-use-lib-env-aliases-from-sveltekit-in-playwright) for me.
Any advancement on the subject ?
@frederikhors have you managed to make it work since ?

**Comment by frederikhors** [comment]
> have you managed to make it work since ?

Nope. We are all waiting for a fix.

**Comment by sowdowdow** [comment]
Workaround **($env only)**
```typescript
import { connect } from "@planetscale/database";
import { drizzle } from "drizzle-orm/planetscale-serverless";
import 'dotenv/config'

var DB_HOST = process.env.DB_HOST
var DB_PASSWORD = process.env.DB_PASSWORD
var DB_USERNAME = process.env.DB_USERNAME

const connection = connect({
  host: DB_HOST,
  username: DB_USERNAME,
  password: DB_PASSWORD,
});

export const db = drizzle(connection);
```

**Comment by oscarhermoso** [comment]
Another workaround, run Playwright in Vitest

https://www.the-koi.com/projects/how-to-run-playwright-within-vitest/

**Comment by mcmxcdev** [comment]
Looks like this became even more relevant now in SvelteKit 2 due to https://kit.svelte.dev/docs/migrating-to-sveltekit-2#generated-tsconfig-json-is-more-strict

**Comment by alexbjorlig** [comment]
I'm running Playwright and SvelteKit `2.0.6`. With the following configuration I have `$lib` and `$tests` working in my playwright test files:

**tests/tsconfig.json**
```ts
{
    "extends": "../tsconfig.json",
    "compilerOptions": {
        "baseUrl": "./",
        "paths": {
            "$lib": ["../src"],
            "$lib/*": ["../src/*"],
            "$tests": ["../tests"],
            "$tests/*": ["../tests/*"]
        }
    }
}

```

**svelte.config.js**
```js
        alias: {
            'tests/*': 'tests/*',
            tests: 'tests',
        },
```

I'm unsure if I actually need the `extend` keyword, but seems to work great.

**Comment by madacol** [comment]
This is relevant https://playwright.dev/docs/test-components#i-have-a-project-that-already-uses-vite-can-i-reuse-the-config

Playwright-for-Svelte uses its own vite internally, so it has a way to inject configs to it

```js
import { defineConfig } from '@playwright/experimental-ct-svelte';

export default defineConfig({
  use: {
    ctViteConfig: {
      // ...
    },
  },
});
```

Keep in mind this is not standard playwright, it uses an official wrapper configured to test Svelte components

**Comment by frederikhors** [comment]
@madacol I cannot make it work with this neither.

**Comment by phcoliveira** [comment]
> > I ended up using a relative import.
> 
> How?
> 
> The issue is when the `$env` or others are imported from the files I import in tests ones.

Well, having a working `$env` is kind of a sub-issue from this one. I have some server-only async functions for fetching data in my application, and I wanted to use these same functions to get such data and determine what to do next in my tests.

For my use case, a relative import still works. However, I do recognise that it does not bring the whole functionality provided by SvelteKit.

**Comment by madacol** [comment]
I did a basic test, and couldn't make it work :disappointed: (didn't try very hard either because I was not interested in testing svelte components)

I ended up making a copy of the lib, and imported dotenv (I'm a fraud :laughing: , but it "works"  :shrug: )

**Comment by frederikhors** [comment]
> a relative import still works

Can you show what do you mean?

**Comment by frederikhors** [comment]
> copy of the lib

Yeah, not very practical when you have hundreds of them.

**Comment by phcoliveira** [comment]
Sorry about that. I forgot that I tried to do what I said, but it did not work, so I reverted to a relative import of the JSON data.

```ts
import {expect, test} from '@playwright/test';
import quizzes from '../../src/lib/game/quiz/private/pt-PT.json' assert {type: 'json'};
// import {getServerQuiz} from '../../src/lib/game/quiz/utils.server';
```

I left this comment so I can remember to try again and find a solution. But, for the moment, I still am replicating part of the functionality provided by the commented function in the test itself.

**Comment by phcoliveira** [comment]
Basically, I need the server data to know which is the right answer to pick in a test of a random quiz.

**Comment by einarpersson** [comment]
Still no fix? I have a utils test folder which I want to be able to import easily everywhere in all tests. Really wish it would "just work"

**Comment by eltigerchino** [comment]
> [WilliamHua/lib-sveltekit-bug-playwright](https://github.com/WilliamHua/lib-sveltekit-bug-playwright)
> 
> It works using version `1.1.1` of SvelteKit, but not `1.1.2`.
> 
> I suspect it's because of this change: #8568

I've forked the reproduction, upgraded the dependencies, and ran `pnpm test`. It seems to recognise the `$lib` alias just fine. I'm going to close this issue but will reopen it if someone can provide a full reproduction. Here's the one I used https://github.com/eltigerchino/8742-sveltekit

In regard to virtual modules such as `$app` or `$env`, an issue already exists for it https://github.com/sveltejs/kit/issues/1485
